`&java.latest` - get latest bell-sw libereca jre

bingling-sama opened the pull request.

bingling-sama committed:
✨ feat(command): new command

Command `&java.latest`
Usage:
&java.latest {type} {version}
type: jdk/jre
version: a correct java feature version number

bingling-sama committed:
Merge branch 'master' of https://github.com/bingling-sama/commspt-bot

SerinaNya commented on models/apis.py:
```diff
@@ -139,3 +140,10 @@ async def get(cls, api_root: str, username: str, texture_type: str):
+
+
+async def getLiberica(version: int, type: str):
+    r = httpx.get(
+        "https://api.bell-sw.com/v1/liberica/releases?version-feature={version}&version-modifier=latest&bitness=64&os=windows&arch=x86&installation-type=installer&bundle-type={type}-full&output=text&fields=downloadUrl"
+    )
+    return r.text
```
- 返回值需要拼接成官网域名的文件下载 URL
- 需要将 query params 单独以 dict 的形式传给 httpx
- 此版本代码中存在 bug，字符串无法被正确格式化

SerinaNya replied:
补充：

需要使用异步调用，请将其改为 async 版本。

<https://www.python-httpx.org/async/>

bingling-sama committed:
🐳 chore(command): improve code

bingling-sama committed:
🐳 chore(command): async support

async support for command `&java.latest`

SerinaNya commented on models/apis.py:
```diff
@@ -139,3 +140,24 @@ async def get(cls, api_root: str, username: str, texture_type: str):
+        file_version = obj["version"]
+        filename = obj["filename"]
+        url = f"https://download.bell-sw.com/java/{file_version}/{filename}"
+        await client.aclose()
+        return url
```
- 删除 L162
- L163 向左取消一个缩进

这是 with 捏。 。

SerinaNya commented on modules/latest_versions.py:
```diff
@@ -25,3 +25,12 @@ async def csl_latest(app: Ariadne, group: Group, mod_loader: str):
+
+@channel.use(CommandSchema('&java.latest {type: str = "jre"} {version: int = "8"}'))
+async def java_latest(app: Ariadne, group: Group, version: int, type: str):
+    url = getLiberica(version, type)
```
无法运行的代码：调用异步函数需要使用 await 关键字捏

bingling-sama committed:
🐳 chore(command): fix bugs

SerinaNya commented on modules/latest_versions.py:
```diff
@@ -25,3 +25,12 @@ async def csl_latest(app: Ariadne, group: Group, mod_loader: str):
     await app.sendGroupMessage(group, MessageChain([Plain(_messages[mod_loader])]))
+
+
+@channel.use(CommandSchema('&java.latest {type: str = "jre"} {version: int = "8"}'))
```
type 和 version 顺序换一下吧

bingling-sama replied:
好的，不过这个命令解析器不会根据输入参数类型判断是哪个参数吗？

bingling-sama committed:
🦄 refactor(command): command &java.latest

change command `&java.latest` params order
Usage:
`&java.latest {version} {type}`
version: a correct java feature version number
type: `jdk`/`jre`
